### ISSUE TITLE
Unclear what happens when multiple PRs call show()

### ISSUE THREAD

**Issue by marcoscaceres** [w3c/browser-payment-api#462]
Unless I missed it, it's unclear in the spec what happens when:

```JS
const pr1 = new PaymentRequest(methods, details, options);
const pr2 = new PaymentRequest(methods, details, options);
p1.show();
p2.show();
```

Should p2 reject? or should it wait till p1 finishes?

**Comment by zkoch** [comment]
I think p2 should reject. @domenic thoughts?

**Comment by domenic** [comment]
Good question. Rejecting seems fine. Alternately you could have some queue of to-be-shown payment requests, but I don't imagine anyone wants to implement that.

Ideally there'd be some way to tell if this was going to reject before you called it, e.g. a `PaymentRequest.canShow` boolean or something. But that's not really necessary.

Here's an important question: what is the scope of this? I think it should be either per browsing context or per top-level browsing context. (It could alternately be per user-agent, i.e. you cannot show a payment request while another tab is showing one. But that seems bad.)

Basically, can iframes do a payment request while the parent document is doing one, or not?

**Comment by rsolomakhin** [comment]
Chrome's implementation currently limits one `.show()` call at a time per browser process. Because the payment process is interactive, this effectively sets a limit of one `.show()` call at a time per user.

**Comment by domenic** [comment]
What if I have two Chrome tabs open? What about if I have two Chrome windows open on separate monitors?

**Comment by rsolomakhin** [comment]
Same browser process in both cases. Is this a spec issue or a UX decision?

**Comment by domenic** [comment]
> Same browser process in both cases.

So, to be clear, what will happen? Will show() reject? Hang the tab and prevent any further JavaScript from running, like alert()? Crash?

>  Is this a spec issue or a UX decision?

Definitely a spec issue, given its observable consequences from code.

**Comment by rsolomakhin** [comment]
The second `.show()` will reject with `AbortError: Request cancelled` in the current implementation. I'm open to improving the error message.

**Comment by marcoscaceres** [comment]
FWIW, ApplePay on Safari also seems to restrict to once per browser instance: it throws if you try to spin up two sessions. It's not clear to me what exception it is throwing tho (in Safari's dev tools, it looks like it's just throwing text :-/... but not sure.). 

You can test it here:
https://stripe.com/docs/apple-pay/web 

Having the ability to only spin up a single payment request at a time seems sensible though. We should standardize on that (Chrome/Safari) behavior. So `AbortError` would be ok, but maybe `InvalidStateError` would be better - because abort really signals that the user aborted.

**Comment by domenic** [comment]
Huh OK, I thought that would be controversial, but if we're all OK with one per user agent then I'm happy to work on speccing that. (EDIT: or someone else; don't block on me!)

I don't feel strongly on the error type. It's a bit weird to say it's an invalid state since the developer can't know ahead of time whether they're going to be screwed over by some other site doing a payment request at the same time, so it's not like they did anything invalid.

**Comment by marcoscaceres** [comment]
Yeah, I don't feel strongly about the error type - just my 2c. 

Thanks for offering to spec it!

**Comment by marcoscaceres** [comment]
FWIW, I think it makes sense to do once per UA because of how payments need to talk to, for instance, watches (or similar hardware). It would be super confusing to know which site is asking for authorization on a watch otherwise.

**Comment by domenic** [comment]
I'm leaning more toward "AbortError" because the site should handle this as if the user aborted. For example, they should show a button to re-launch the process. Whereas "InvalidStateError"s usually mean a logic bug in the application that should not be caught and reacted to (besides maybe logging it to the server).

**Comment by marcoscaceres** [comment]
Reopening... it might be that Apple's implementation changed (?)... the single payment sheet per browser, but now if site A is showing a sheet, and site b calls `.show()`, site A aborts, then site B shows a payment sheet. 

Basically, you can play payment sheet whack-a-mole. 

If the above is confusing, I can make a video showing etsy and shopify's playground.

**Comment by marcoscaceres** [comment]
We need to add a note to the spec that we are still trying to figure out what the right behavior is (or that UAs might vary here). Gecko is currently following the spec, but our UX folks are investigating different options.

**Comment by marcoscaceres** [comment]
Closing, as the original issue was addressed in #487.
